### PR TITLE
Add suport for TransactionNext

### PR DESCRIPTION
**Pull request by popenta**

### CODE REVIEW

File: src/providers.dev.net.spec.ts [hunk 1]
```diff
@@ -315,5 +315,51 @@ describe("test network providers on devnet: Proxy and API", function () {
         assert.deepEqual(apiResponse.logs.events[0].additionalData, [TransactionEventData.fromBase64("dGVzdA==")]);
         assert.deepEqual(proxyResponse.logs.events[0].additionalData, [TransactionEventData.fromBase64("dGVzdA==")]);
     });
-});
 
+    it.only("should send both `Transaction` and `TransactionNext`", async function () {
```
Comment (andreibancioiu) [inline src/providers.dev.net.spec.ts:319]: Without `only`.

File: src/providers.dev.net.spec.ts [hunk 2]
```diff
@@ -315,5 +315,51 @@ describe("test network providers on devnet: Proxy and API", function () {
...
+            nonce: BigInt(8),
+            value: BigInt(0),
+            receiver: "erd1qyu5wthldzr8wx5c9ucg8kjagg0jfs53s8nr3zpz3hypefsdd8ssycr6th",
+            sender: "erd1zztjf9fhwvuvquzsllknq4qcmffwad6n0hjtn5dyzytr5tgz7uas0mkgrq",
+            data: new Uint8Array(Buffer.from("test")),
+            gasPrice: BigInt(1000000000),
+            gasLimit: BigInt(80000),
+            chainID: "D",
+            version: 2,
+            signature: new Uint8Array(Buffer.from("3fa42d97b4f85442850340a11411a3cbd63885e06ff3f84c7a75d0ef59c780f7a18aa4f331cf460300bc8bd99352aea10b7c3bc17e40287337ae9f9842470205", "hex")),
```
Comment (andreibancioiu) [inline src/providers.dev.net.spec.ts:348]: Can be `Buffer.from(...)` without `new Uint8Array()`.

File: src/providers.dev.net.spec.ts [hunk 3]
```diff
@@ -315,5 +315,51 @@ describe("test network providers on devnet: Proxy and API", function () {
...
+            guardian: "",
+            guardianSignature: new Uint8Array(),
+            options: 0
+        }
+
+        const apiLegacyTxHash = await apiProvider.sendTransaction(transaction);
+        const apiTxNextHash = await apiProvider.sendTransaction(transactionNext);
+
+        const proxyLegacyTxHash = await apiProvider.sendTransaction(transaction);
+        const proxyTxNextHash = await apiProvider.sendTransaction(transactionNext);
```
Comment (andreibancioiu) [inline src/providers.dev.net.spec.ts:360]: Using `proxyProvider.sendTransaction()`.

File: src/providers.dev.net.spec.ts [hunk 4]
```diff
@@ -315,5 +315,51 @@ describe("test network providers on devnet: Proxy and API", function () {
...
+
+        assert.equal(apiLegacyTxHash, proxyLegacyTxHash);
+        assert.equal(apiTxNextHash, proxyTxNextHash);
```
Comment (andreibancioiu) [inline src/providers.dev.net.spec.ts:363]: Correct, if we keep the test sender private, the transactions will always be accepted (but never executed, due to high nonce) :+1:

File: src/proxyNetworkProvider.ts [hunk 5]
```diff
@@ -122,13 +122,25 @@ export class ProxyNetworkProvider implements INetworkProvider {
...
+        const transaction = prepareTransactionNextForBroadcasting(tx);
+        const response = await this.doPostGeneric("transactions", transaction);
         return response.txHash;
     }
 
-    async sendTransactions(txs: ITransaction[]): Promise<string[]> {
-        const data: any = txs.map(tx => tx.toSendable());
+    async sendTransactions(txs: ITransaction[] | ITransactionNext[]): Promise<string[]> {
+        let data: any[];
+        if ("toSendable" in txs[0]){
```
Comment (andreibancioiu) [inline src/proxyNetworkProvider.ts:138]: This fails if the array is empty. Not a very common case, but I think we should handle it.

I think it's also good to apply the idea you've mentioned on Friday - accept array of `legacy | next` - would result in a nicer code, as well.

File: src/transactions.ts [hunk 6]
```diff
@@ -1,10 +1,30 @@
 import { TransactionStatus } from "./transactionStatus";
 import { ContractResults } from "./contractResults";
 import { Address } from "./primitives";
-import { IAddress } from "./interface";
+import { IAddress, ITransactionNext } from "./interface";
 import { TransactionLogs } from "./transactionLogs";
 import { TransactionReceipt } from "./transactionReceipt";
 
+export function prepareTransactionNextForBroadcasting(transaction: ITransactionNext): any {
```
Comment (andreibancioiu) [inline src/transactions.ts:8]: Perhaps also do a `perpareTransactionForBroadcasting(legacy | next)` and have here (only in one place) the check about having "toSendable"?

File: src/proxyNetworkProvider.ts [hunk 7]
```diff
@@ -139,8 +151,14 @@ export class ProxyNetworkProvider implements INetworkProvider {
...
 
-    async simulateTransaction(tx: ITransaction): Promise<any> {
-        let response = await this.doPostGeneric("transaction/simulate", tx.toSendable());
+    async simulateTransaction(tx: ITransaction | ITransactionNext): Promise<any> {
+        if ("toSendable" in tx){
+            const response = await this.doPostGeneric("transaction/simulate", tx.toSendable());
+            return response;
+        }
+        
+        const transaction = prepareTransactionNextForBroadcasting(tx);
```
Comment (andreibancioiu) [inline src/proxyNetworkProvider.ts:160]: A comment below - perhaps have the preparation logic extracted in `transactions.ts` - to handle both, not only "next".

File: src/providers.dev.net.spec.ts (hunk 1 above)
Comment (popenta) [reply src/providers.dev.net.spec.ts:319]: Removed. Thanks!

File: src/providers.dev.net.spec.ts (hunk 2 above)
Comment (popenta) [reply src/providers.dev.net.spec.ts:348]: Done!

File: src/providers.dev.net.spec.ts (hunk 3 above)
Comment (popenta) [reply src/providers.dev.net.spec.ts:360]: Changed!

File: src/providers.dev.net.spec.ts (hunk 4 above)
Comment (popenta) [reply src/providers.dev.net.spec.ts:363]: Indeed.

File: src/proxyNetworkProvider.ts (hunk 7 above)
Comment (popenta) [reply src/proxyNetworkProvider.ts:160]: Done!

File: src/proxyNetworkProvider.ts (hunk 5 above)
Comment (popenta) [reply src/proxyNetworkProvider.ts:138]: Now accepting `(ITransaction | ITransactionNext)[]`

File: src/transactions.ts (hunk 6 above)
Comment (popenta) [reply src/transactions.ts]: Done

**Review by andreibancioiu — APPROVED** [review]

**Review by bogdan-rosianu — APPROVED** [review]

**Pull request merged**